### PR TITLE
Bug 1433475 — Keyboard — enable up and down arrows in location bar.

### PR DESCRIPTION
**Pull request by jhugman**
This PR allows the user to select URL completions (from bookmarks and history) to be selected via the keyboard. 

Desktop leaves the cursor in the location field, which means than horizontal cursor movement is possible.

We aren't able to do this, given that the search suggestions for the query appear horizontally.

Please file follow up usability bugs.

r? for @farhan.

https://bugzilla.mozilla.org/show_bug.cgi?id=1433475

### CODE REVIEW

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 1]
```diff
@@ -1753,6 +1753,7 @@ extension BrowserViewController: SearchViewControllerDelegate {
     }
 
     func searchViewControllerDidFinishHighlighting(_ searchViewController: SearchViewController) {
+        searchViewController.resignFirstResponder()
     }
```
Comment (jhugman) [inline Client/Frontend/Browser/BrowserViewController.swift]: This is called when the user has typed DOWN, then UP — i.e. focus should move back to the location bar.

File: Client/Frontend/Browser/SearchViewController.swift [hunk 2]
```diff
@@ -420,6 +426,87 @@ class SearchViewController: SiteTableViewController, KeyboardHelperDelegate, Loa
...
+}
+
+extension SearchViewController {
+
+    override var canBecomeFirstResponder: Bool {
+        return true
+    }
+
+    func handleKeyCommands(sender: UIKeyCommand) {
+        let initialSection = SearchListSection.bookmarksAndHistory.rawValue
```
Comment (jhugman) [inline Client/Frontend/Browser/SearchViewController.swift]: We might change this value when we get we get to [Bug 1433476][1].

[1]: https://bugzilla.mozilla.org/show_bug.cgi?id=1433476

File: Client/Frontend/Browser/SearchViewController.swift [hunk 3]
```diff
@@ -420,6 +426,87 @@ class SearchViewController: SiteTableViewController, KeyboardHelperDelegate, Loa
...
+        self.tableView(tableView, didHighlightRowAt: next)
+        tableView.selectRow(at: next, animated: false, scrollPosition: .middle)
+    }
+
+    func didSelectTableCellKeyCommand() {
+        guard let current = tableView.indexPathForSelectedRow else {
+            return
+        }
+
+        tableView(tableView, didSelectRowAtIndexPath: current)
```
Comment (jhugman) [inline Client/Frontend/Browser/SearchViewController.swift]: Calling this delegate here feels weird. How do we perform the same action a tap performs?

File: Client/Frontend/Browser/SearchViewController.swift [hunk 4]
```diff
@@ -420,6 +426,87 @@ class SearchViewController: SiteTableViewController, KeyboardHelperDelegate, Loa
...
+                } else {
+                    nextSection += 1
+                    nextItem = 0
+                }
+            } else {
+                nextItem += 1
+            }
+        }
+        let next = IndexPath(item: nextItem, section: nextSection)
+        self.tableView(tableView, didHighlightRowAt: next)
```
Comment (jhugman) [inline Client/Frontend/Browser/SearchViewController.swift]: Calling this delegate here feels weird. `tableView.selectRow` does the highlighting.

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 5]
```diff
@@ -1747,6 +1747,14 @@ extension BrowserViewController: SearchViewControllerDelegate {
 
         self.present(navController, animated: true, completion: nil)
     }
+
+    func searchViewController(_ searchViewController: SearchViewController, didHighlightText text: String, search: Bool) {
+        self.urlBar.setLocation(text, search: search)
+    }
+
+    func searchViewControllerDidFinishHighlighting(_ searchViewController: SearchViewController) {
+        searchViewController.resignFirstResponder()
```
Comment (jhugman) [inline Client/Frontend/Browser/BrowserViewController.swift]: This is called when the user has typed DOWN, then UP — i.e. focus should move back to the location bar.

Also tried: explicitly setting BVC or the location bar as first responder.

**Comment by farhanpatel** [comment]
I think to simplify this. It might be easier to never change the focus from the AutocompleteTextField. Instead just pass the up/down keypress from the textfield down to the SearchViewController. That way you dont have to worry about show/hiding the keyboard. 

Where this gets tricky is handling the search suggestions. Because they are shown horizontally, tapping left/right to switch between them wouldnt work because on the textfield that would move your cursor. I think this is why safari/chrome/firefox desktop show autocomplete suggestions as a list instead of the way we do.

**Comment by jhugman** [comment]
> It might be easier to never change the focus from the AutocompleteTextField.
> Because they are shown horizontally, tapping left/right to switch between them wouldnt work because on the textfield that would move your cursor.

Yeah, I think you've answered your own question about why we can't do this. 

> That way you dont have to worry about show/hiding the keyboard.
Ug. I wasn't considering the soft keyboard for `UIKeyCodes`.

**Comment by jhugman** [comment]
Ok, I think I've cracked it. 

Looking at desktop, they seem to be leaving the cursor with the text view, and you can move it around with the cursor.

I'm not altogether happy with what we've got here — but we're limited by the design of the search auto-complete. I'm leaving the horizontal keys for the next follow up bug.

**Comment by farhanpatel** [comment]
I'm getting a pretty bad crash that happens when there are no urls in the searchsuggestions tableview. 
1. Type until there are no url suggestions 
2. press up/down on the keyboard
3. crashes 😢 

I think in the optimal case I dont think we should move focus away from the keyboard. I think this should work exactly like Desktop. For example. if you want to go to `reddit.com/r/ios` and you already have `reddit.com` in your history. 
you should be able to type `re` arrow down to `reddit.com` and then press right array to get to the end of the url where you can continue typing `/r/ios`

right now. if you want to continue typing for any reason you have to tap the urlbar again.

File: Client/Frontend/Browser/SearchViewController.swift [hunk 6]
```diff
@@ -69,6 +71,10 @@ class SearchViewController: SiteTableViewController, KeyboardHelperDelegate, Loa
 
     static var userAgent: String?
 
+    var isCellHighlighted: Bool {
```
Comment (farhanpatel) [inline Client/Frontend/Browser/SearchViewController.swift]: I dont think this is used anywhere.

**Pull request merged**